### PR TITLE
Region properties middle key

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
Add 2 middle_key in region properties response so that we can use the key to split region at approximate middle.

## What are the type of the changes? (mandatory)
Improvement.

## How has this PR been tested? (mandatory)
make dev.

## Add a few positive/negative examples (optional)
tikv-ctl output changes to:
```
$ target/debug/tikv-ctl --host localhost:12347 region-properties -r 4
num_files: 1
num_entries: 3348
num_deletes: 2225
mvcc.min_ts: 401128769676902406
mvcc.max_ts: 401144011554816001
mvcc.num_rows: 1079
mvcc.num_puts: 1103
mvcc.num_versions: 1123
mvcc.max_row_versions: 21
middle_key: zmBootstr\377a\377pKey\000\000\377\000\000\373\000\000\000\000\000\377\000\000s\000\000\000\000\000\372\372n\347\2020\037\377\372
```

### CODE REVIEW

**Comment by hicqu** [comment]
ping @huachaohuang PTAL.

File: src/raftstore/store/util.rs [hunk 1]
```diff
@@ -326,6 +327,37 @@ pub fn get_region_approximate_size_cf(
     Ok(size)
 }
 
+/// Get the approxmiate middle key of the region. The returned key maybe
+/// is timpstamped if transaction KV is used, and must start with "z".
```
Comment (huachaohuang) [inline src/raftstore/store/util.rs]: s/timpstamped/timestamped

File: src/raftstore/store/util.rs [hunk 2]
```diff
@@ -326,6 +327,37 @@ pub fn get_region_approximate_size_cf(
...
+    let range = Range::new(&start, &end);
+    let collection = db.get_properties_of_tables_in_range(cf, &[range])?;
+
+    let mut keys = Vec::new();
+    for (_, v) in &*collection {
+        let props = SizeProperties::decode(v.user_collected_properties())?;
+        keys.extend(
+            props
+                .index_handles
+                .range::<[u8], _>((Excluded(start.as_slice()), Excluded(end.as_slice())))
```
Comment (Connor1996) [inline src/raftstore/store/util.rs]: why both are excluded?

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/raftstore/store/util.rs (hunk 2 above)
Comment (hicqu) [reply src/raftstore/store/util.rs]: The middle key is supposed to be used for split regions. So we don't hope the key equals to region's start key or end key.

File: src/raftstore/store/util.rs [hunk 3]
```diff
@@ -326,6 +327,37 @@ pub fn get_region_approximate_size_cf(
     Ok(size)
 }
 
+/// Get the approxmiate middle key of the region. The returned key maybe
+/// is timestamped if transaction KV is used, and must start with "z".
+pub fn get_region_approximate_middle_cf(
```
Comment (AndreMouche) [inline src/raftstore/store/util.rs]: Should we point out what does `middle` means?

File: src/raftstore/store/util.rs [hunk 4]
```diff
@@ -326,6 +327,37 @@ pub fn get_region_approximate_size_cf(
...
+                .index_handles
+                .range::<[u8], _>((Excluded(start.as_slice()), Excluded(end.as_slice())))
+                .map(|(k, _)| k.to_owned()),
+        );
+    }
+    keys.sort();
+    if keys.is_empty() {
+        return Ok(None);
+    }
+    let middle = (keys.len() - 1) / 2;
```
Comment (AndreMouche) [inline src/raftstore/store/util.rs]: why sub 1?

Comment (AndreMouche) [reply src/raftstore/store/util.rs]: got it, but maybe we could add a comment here?　@hicqu

**Pull request merged**